docs/config.md
# Brevia Configuration

There are two type of configuration in a Brevia project:

* general configuration set through environment variables or a simple `.env` file, [see below](#configuration-customization)
* specific configuration of a [single `collection`](collection_config.md) mainly for RAG applications, that are stored in the collection metadata

In this page we will explain the general configuration items in Brevia.

All configuration items have working defaults with the exception of:

* external services settings like LLM APIs or monitor/debugging tools (like Langsmith). Those services rely on specific environment variables that we cannot include or predict here
* database connection configuration

## Configuration customization

General configurations can be customized in various ways, some modes have precedence over others.

1. through a `.env` file in the app root folder
1. through environment variables
1. through the [`config` table records](database.md#schema) of the database

Each of the described modes prevails over the previouse ones in the order in which they are listed (DB configuration wins).

Database configuration can be done through [configuration endpoints](endpoints_overview.md#configuration-endpoints) but has some limitations: you cannot change some parameters such as the DB connection or security tokens.

## Database connection

These settings define the connection to your Postgres database with `pgvector` extension:

* `PGVECTOR_DRIVER`: `psycopg2` default should be left unchanged, you may want to avoid this item in your configuration
* `PGVECTOR_HOST`: the database host name or address, `localhost` as default
* `PGVECTOR_PORT`: the database connection port, `5432` as default
* `PGVECTOR_DATABASE`: the database name, `brevia` as default
* `PGVECTOR_USER`: database connection user, no default
* `PGVECTOR_PASSWORD`: database connection password, no default
* `PGVECTOR_POOL_SIZE`: connection pool size, `10` as default

If you prefer to define a single DSN URI for your connection you can use `PGVECTOR_DSN_URI` env variable. If this variable is set the `PGVECTOR_*` variables above will be ignored.
An example of such uri using Brevia defaults can look like `postgresql+psycopg2://user:password@localhost:5432/brevia`

## External services

As example of external services here we include OpenAI, Cohere, Anthropic, DeepSeek and LangSmith but it may include any LLM supported by LangChain. Those services usually need some specific env variables like API keys or tokens. If you don't set these variables, the services that depend on them may not work properly.

By using [`BREVIA_ENV_SECRETS` var as explained below](#brevia-env-secrets) you make sure that these secrets will be available as environment variables.

### Model providers

Brevia supports vistually any LLM provider, each with its own configuration. The following environment variables are necessary to configure them:

* to use OpenAI models via API you need to have a valid API KEY and define an `OPENAI_API_KEY` env var with its value. Please refer to [OpenAI model](models/openai.md) page for further details on using OpenAI.
* similarly to use Cohere model via API you need a valid API KEY that must be defined in `COHERE_API_KEY` env var. Have a look at [Cohere model](models/cohere.md) for more integration details.
* to use Anthropic models via API you need a valid API KEY that must be defined in `ANTHROPIC_API_KEY` env var. Have a look at [Anthropic model](models/anthropic.md) for more details.
* to use DeepSeek models via API you need a valid API KEY that must be defined in `DEEPSEEK_API_KEY` env var. Have a look at [DeepSeek model](models/deepseek.md) for more details.

See [Providers](#providers) paragraph below on available providers, models and related environment variables.
### LangSmith

If you want to use [LangSmith](https://www.langchain.com/langsmith) to monitor your LLM application with Brevia you need to set these variables:

* `LANGCHAIN_TRACING_V2`: set to `True`, or any string actually
* `LANGCHAIN_ENDPOINT`: the endpoint used, like `https://api.smith.langchain.com`
* `LANGCHAIN_API_KEY`: your LangSmith API KEY
* `LANGCHAIN_PROJECT`: the name of your Brevia project

### Brevia env secrets

As mentioned above to make these variables available as environment variables, if not done explicitly outside Brevia, you can use a special JSON object variable `BREVIA_ENV_SECRETS` in your `.env` file where each key/value pair will be loaded in the environment at startup.

This variable could be something like `BREVIA_ENV_SECRETS='{"OPENAI_API_KEY": "#########", "COHERE_API_KEY": "#########"}'`

## Security

To enable a basic security support via tokens you may want to set the following variables:

* `TOKENS_SECRET`: the secret used when enconding and decoding tokens
* `TOKENS_USERS`: an optional comma separated list of valid user names that must be present in a token
* `STATUS_TOKEN`: a special token to be used only by the `/status` endpoint from 3rd party monitoring tools

Please check the [Security](security.md) section for more details on security support on Brevia.

## Index and search

This section focuses on the crucial components of indexing and searching for relevant information within your knowledge base.

### Embeddings

This variable specifies the type of embedding model used to convert text documents into numerical vectors. In this case, it's set to `openai-embeddings`, indicating that you'll be using OpenAI's embedding service.

`EMBEDDINGS='{"_type": "openai-embeddings"}'`

### Supported Embedding Services

`openai-embeddings`: Utilize OpenAI's embedding service for efficient conversion of text to numerical representations.

`cohere-embeddings`: Leverage Cohere's embedding service for alternative embedding calculations.

See [other models](models/other_models.md) for integrations with other 3rd party models

### Text Segmentation

`TEXT_CHUNK_SIZE`
This variable controls the maximum size of individual text chunks during processing. Large documents are split into smaller segments for efficient handling by the embedding model.

Default Value: 2000 (token)

Adjust this value based on your document sizes and hardware resources. Larger chunks typically yield more accurate embeddings, but require more memory. Experiment to find the optimal balance for your setup.

`TEXT_CHUNK_OVERLAP`
This variable specifies the amount of overlap between consecutive text chunks. Overlap ensures continuity within the document and helps capture contextual information across sections.

Default Value: 100 (token)

Increase the overlap for documents with important cross-sectional references, but reduce it for faster processing of independent sections.
Consider experimenting with different values based on your document characteristics.

example:

```bash
TEXT_CHUNK_SIZE=2000
TEXT_CHUNK_OVERLAP=100
```

`TEXT_SPLITTER`
This variable is an optional JSON string configuration, used to override the default text splitter

Can be something like:

`'{"splitter": "my_project.CustomSplitter", "some_var": "some_value"}'`

Where:

* `splitter` key must be present and point to a module path of a valid
   retriever class extending langchain `TextSplitter`
* other splitter constructor attributes can be specified in the configuration,
    like `some_var` in the above example

## Q&A and Chat

Under the hood of Q&A and Chat actions (see [Chat and Search](chat_search.md) section) you can configure models and behaviors via these variables:

* `QA_COMPLETION_LLM`: configuration for the main conversational model, used by `/chat` and `/completion` endpoints; a JSON string is used to configure the corresponding LangChain chat model class; an OpenAI instance is used as default: `'{"model_provider": "openai", "model": "gpt-4o-mini", "temperature": 0, "max_tokens": 2000}'` where for instance `model` and other attributes can be adjusted to meet your needs
* `QA_FOLLOWUP_LLM`: configuration for the follow-up question model, used by `/chat` endpoint defining a follow up question for a conversation usgin chat history; a JSON string; an OpenAI instance used as default `'{"model_provider": "openai", "model": "gpt-4o-mini", "temperature": 0, "max_tokens": 500}'`
* `QA_FOLLOWUP_SIM_THRESHOLD`: a numeric value between 0 and 1 indicating similarity threshold between questions to determine if chat history should be used, defaults to `0.735`
* `QA_NO_CHAT_HISTORY`: disables chat history entirely if set to `True` or any other value
* `SEARCH_DOCS_NUM`: default number of documents used to search for answers, defaults to `4`
* `QA_RETRIEVER`: optional configuration for a custom retriever class, used by `/chat`  endpoint, it's a JSON string defining a custom class and optional attributes; an example configuration can be `'{"retriever": "my_project.CustomRetriever", "some_var": "some_value"}'` where `retriever` key must be present with a module path pointing to a valid retriever class extending langchain `BaseRetriever` whereas other constructor attributes can be specified in the configuration, like `some_var` in the above example

## Summarization

To configure summarize related actions in `/summarize` or `/upload_summarize` endpoints the related environment variables are:

* `SUMMARIZE_LLM`: the LLM to be used, a JSON string using the same format of `QA_COMPLETION_LLM` in the above paragraph; defatults to an OpenAI instance `'{"model_provider": "openai", "model": "gpt-4o-mini" "temperature": 0, "max_tokens": 2000}'`
* `SUMM_TOKEN_SPLITTER`: the maximum size of individual text chunks processed during summarization, defaults to `4000` - see `TEXT_CHUNK_SIZE` in [Text Segmentation](#text-segmentation) paragraph
* `SUMM_TOKEN_OVERLAP`: the amount of overlap between consecutive text chunks, defaults to `500` - see `TEXT_CHUNK_OVERLAP` in [Text Segmentation](#text-segmentation) paragraph
* `SUMM_DEFAULT_CHAIN`: chain type to be used if not specified, defaults to `stuff`

## Providers

A list of supported providers and its models can be defined via `PROVIDERS` variable. This variable is a JSON string with the following format:

```JSON
[
    {
        "model_provider": "openai",
        "models": [
            {
                "name": "o1-mini"
            },
            {
                "name": "o3-mini"
            },
            {
                "name": "gpt-4o"
            },
            {
                "name": "gpt-4o-mini"
            }
        ]
    },
    {
        "model_provider": "cohere",
        "models": [
            {
                "name": "command-r",
                "tokens_limit": 128000
            },
            {
                "name": "command-r-plus",
                "tokens_limit": 128000
            }
        ]
    },
    {
        "model_provider": "anthropic",
        "models": [
            {
                "name": "claude-3-7-sonnet-20250219"
            },
            {
                "name": "claude-3-5-sonnet-20241022"
            }
        ]
    },
    {
        "model_provider": "ollama",
        "models": [
            {
                "name": "llama3.2:latest"
            }
        ]
    },
    {
        "model_provider": "deepseek",
        "models": [
            {
                "name": "deepseek-chat"
            },
            {
                "name": "deepseek-reasoner"
            }
        ]
    }
]
```

If not set this variable will be automatically populated with the current available providers and models at Brevia startup with the content of [the `/providers` endpoint response](endpoints_overview.md#providers-endpoints).

To help with setup of external providers via UI a `PROVIDERS_ENV_VARS` JSON configuration can be used to list knwon environment variable that con be used for every provider.
An example of this configuration is:

```JSON
{
    "openai": ["OPENAI_API_KEY","OPENAI_ORG_ID","OPENAI_API_BASE"],
    "anthropic": ["ANTHROPIC_API_KEY"],
    "cohere": ["COHERE_API_KEY"],
    "deepseek": ["DEEPSEEK_API_KEY"],
    "ollama": ["OLLAMA_HOST"]
}
```


docs/collection_config.md
# Collection Configuration

Specific configuration options for a single `collection` can be set in Brevia. Those options are stored as JSON in `langchain_pg_collection.cmetadata` database column and are used mainly for RAG applications.

This way you can also override some [global configurations](config.md) on a specific `collection` to use, for instance, different LLM settings or embeddings/indexing strategies.

All configuration items described here below can be set as keys in `cmetadata` JSON field of a single collection using [`PATCH /collections/{{collection_id}}` endpoint](endpoints_overview.md#collections-endpoints)

## Prompt

- `prompts` (JSON) configuration of the custom prompts used in the [Q&A & chat](chat_search.md) actions, see [Prompt Management](prompt_management.md) for more details.

## Index and search

### Embeddings

- `embeddings` (JSON) configuration for the embedding engins - overrides [`EMBEDDINGS`](config.md#embeddings)

### Text Segmentation

- `chunk_size` (integer) maximum size of individual text chunks in documents index - overrides [``TEXT_CHUNK_SIZE`](config.md#text-segmentation)
- `chunk_overlap` (integer) amount of overlap between consecutive text chunk - overrides [``TEXT_CHUNK_OVERLAP`](config.md#text-segmentation)
- `text_splitter` (JSON) configuration of a custom text splitter - overrides [``TEXT_CHUNK_OVERLAP`](config.md#text-segmentation)

### Search

- `docs_num` (integer) the number of documents to extract in a search or Q&A action - will override [`SEARCH_DOCS_NUM`](config.md#qa-and-chat)

## Q&A and Chat

You can configure models and behaviors of [Q&A and chat actions](chat_search.md) via these variables:

- `qa_completion_llm` (JSON) configuration for the main conversational model - overrides [`QA_COMPLETION_LLM`](config.md#qa-and-chat)
- `qa_followup_llm` (JSON) configuration for the follow-up question model - overrides [`QA_FOLLOWUP_LLM`](config.md#qa-and-chat)
- `qa_retriever` (JSON) configuration for a custom retriever class - overrides [`QA_RETRIEVER`](config.md#qa-and-chat)

## Documents

There are some document related configurations that are collection-specific and are not present in the global configurations.

### Metadata

- `documents_metadata` (JSON) JSON Schema of the documents metadata
- `metadata_defaults` (JSON) default values of the documents metadata, if any, divided by document type

### Load Options

- `file_upload_options` (JSON) custom options when loading files, for instance OCR options for PDFs
- `link_load_options` (JSON) custom options when loading links, for instance selector rules to use


brevia/settings.py
"""Settings module"""
import logging
from functools import lru_cache
from typing import Annotated, Any
from os import environ, path
from urllib import parse
from sqlalchemy import NullPool, create_engine, Column, String, func, inspect
from sqlalchemy.engine import Connection
from sqlalchemy.dialects.postgresql import TIMESTAMP
from sqlalchemy.orm import Session
from langchain_community.vectorstores.pgembedding import BaseModel
from langchain.globals import set_verbose
from pydantic import Field, Json, PrivateAttr, ValidationError
from pydantic_settings import BaseSettings, SettingsConfigDict


class Settings(BaseSettings):
    """Brevia settings"""
    model_config = SettingsConfigDict(
        env_file='.env', extra='ignore'
    )
    _defaults: dict[str, Any] = PrivateAttr(default={})

    verbose_mode: bool = False

    # Postgres+pgvector vector db
    pgvector_host: str = Field(default='localhost', exclude=True)
    pgvector_driver: str = Field(default='psycopg2', exclude=True)
    pgvector_port: int = Field(default=5432, exclude=True)
    pgvector_database: str = Field(default='brevia', exclude=True)
    pgvector_user: str = Field(default='', exclude=True)
    pgvector_password: str = Field(default='', exclude=True)
    pgvector_pool_size: int = Field(default=10, exclude=True)
    # optional DSN URI, if set other pgvector_* settings are ignored
    pgvector_dsn_uri: str = Field(default='', exclude=True)

    # Tokens
    tokens_secret: str = Field(default='', exclude=True)
    tokens_users: str = Field(default='', exclude=True)
    status_token: str = Field(default='', exclude=True)

    # External services API keys (deprecated)
    # You should use `brevia_env_secrets` to store secrets that
    # must be available as environment variables
    openai_api_key: Annotated[
        str,
        Field(deprecated='Use `brevia_env_secrets` instead')
    ] = Field(default='', exclude=True)
    cohere_api_key: Annotated[
        str,
        Field(deprecated='Use `brevia_env_secrets` instead')
    ] = Field(default='', exclude=True)

    # Test models - only in unit tests
    use_test_models: bool = Field(default=False, exclude=True)

    # Index - text splitter settings
    text_chunk_size: int = 2000
    text_chunk_overlap: int = 200
    text_splitter: Json[dict[str, Any]] = '{}'  # custom splitter settings

    # Search
    search_docs_num: int = 4

    # LLM settings
    qa_completion_llm: Json[dict[str, Any]] = """{
        "model": "gpt-4o-mini",
        "model_provider": "openai",
        "temperature": 0,
        "max_tokens": 2000
    }"""
    qa_followup_llm: Json[dict[str, Any]] = """{
        "model": "gpt-4o-mini",
        "model_provider": "openai",
        "temperature": 0,
        "max_tokens": 500
    }"""
    summarize_llm: Json[dict[str, Any]] = """{
        "model": "gpt-4o-mini",
        "model_provider": "openai",
        "temperature": 0,
        "max_tokens": 2000
    }"""

    # Environment secrets, dictionary of variables that must be present as environment
    # variables
    # (Pydantic throws an error using SecretStr as dict value type)
    brevia_env_secrets: Json[dict[str, str]] = '{}'

    # Embeddings
    embeddings: Json = '{"_type": "openai-embeddings"}'
    # every vendor has its own embeddings vector size, some vendors have multiple sizes
    # this value should be unique in your project to avoid calculation errors
    embeddings_size: Annotated[
        int,
        Field(deprecated='Define size in `embeddings` conf')
    ] = Field(default=1536, exclude=True)

    # QA
    qa_no_chat_history: bool = False  # don't load chat history
    qa_followup_sim_threshold: float = 0.735  # similitude threshold in followup
    qa_retriever: Json[dict[str, Any]] = '{}'  # custom retriever settings

    # Summarization
    summ_default_chain: str = 'stuff'
    summ_token_splitter: int = 4000
    summ_token_overlap: int = 500

    # Prompt files base path - local or shared file system
    prompts_base_path: str = Field(
        default=f'{path.dirname(__file__)}/prompts',
        exclude=True,
    )

    # Providers: list of available providers and models
    providers: Json = '[]'

    # List of known providers environment variables that could be used by
    # external services (mainly LLM providers)
    providers_env_vars: Json[dict[str, list[str]]] = """{
        "openai": ["OPENAI_API_KEY","OPENAI_ORG_ID","OPENAI_API_BASE"],
        "anthropic": ["ANTHROPIC_API_KEY"],
        "cohere": ["COHERE_API_KEY"],
        "deepseek": ["DEEPSEEK_API_KEY"],
        "ollama": ["OLLAMA_HOST"]
    }"""

    # App metadata
    block_openapi_urls: bool = False

    def update(
        self,
        other: dict[str, Any] | BaseSettings,
    ) -> None:
        """Update settings fields, used when updating from db and in unit tests"""
        keys = Settings.model_fields.keys()
        if not isinstance(other, BaseSettings):
            other = {k.lower(): v for k, v in other.items() if hasattr(self, k)}
            keys = other.keys()
            other = Settings(**other)
        for key in keys:
            newattr = getattr(other, key)
            if key == 'brevia_env_secrets':
                current_secrets = getattr(self, key)
                newattr = {**current_secrets, **newattr}
            setattr(self, key, newattr)

    def setup_environment(self):
        """Setup some useful environment variables from `brevia_env_secrets`"""
        log = logging.getLogger(__name__)
        for key, value in self.brevia_env_secrets.items():
            environ[key] = value
            log.info('"%s" env var set', key)
        for provider in self.providers_env_vars:
            for key in self.providers_env_vars[provider]:
                if key in environ and key not in self.brevia_env_secrets.keys():
                    self.brevia_env_secrets[key] = environ[key]
                    log.info('"%s" added to env secrets', key)

    def connection_string(self) -> str:
        """ Db connection string from Settings """
        if self.pgvector_dsn_uri:
            return self.pgvector_dsn_uri

        driver = self.pgvector_driver
        host = self.pgvector_host
        port = self.pgvector_port
        database = self.pgvector_database
        user = self.pgvector_user
        password = parse.quote_plus(self.pgvector_password)

        return f"postgresql+{driver}://{user}:{password}@{host}:{port}/{database}"

    def update_from_db(self):
        """Update settings from db"""
        try:
            engine = create_engine(self.connection_string(), poolclass=NullPool)
            insp = inspect(engine)
            if insp.has_table(ConfigStore.__tablename__):
                db_conf = read_db_conf(engine.connect())
                self.update(db_conf)
            engine.dispose()
        except Exception as exc:  # pylint: disable=broad-exception-caught
            logging.getLogger(__name__).error('Failed to read config from db: %s', exc)

    def setup_defaults(self):
        """Setup defaults"""
        self._defaults = self.model_dump()

    def get_configurable_schema(self) -> dict[str, Any]:
        """Get JSON Schema of configurable properties, with actual defaults"""
        schema = self.model_json_schema(mode='serialization')
        for key, value in schema['properties'].items():
            value['default'] = self._defaults[key]

        return schema


class ConfigStore(BaseModel):
    # pylint: disable=too-few-public-methods,not-callable
    """ Config table """
    __tablename__ = "config"

    config_key = Column(String(), nullable=False, unique=True)
    config_val = Column(String(), nullable=False)
    created = Column(
        TIMESTAMP(timezone=False),
        nullable=False,
        server_default=func.current_timestamp(),
    )
    modified = Column(
        TIMESTAMP(timezone=False),
        nullable=False,
        server_default=func.current_timestamp(),
        onupdate=func.current_timestamp(),
    )


@lru_cache
def get_configurable_keys():
    """Get configurable keys"""
    return Settings().model_dump().keys()


def read_db_conf(connection: Connection) -> dict[str, str]:
    """ Read all config records """
    with Session(connection) as session:
        query = session.query(ConfigStore.config_key, ConfigStore.config_val)
        # Filter out non-configurable settings
        keys = get_configurable_keys()
        return {key: value for key, value in query.all() if key in keys}


def update_db_conf(connection: Connection, items: dict[str, str]) -> dict[str, str]:
    """ Update some config items """
    # Filter out non-configurable settings
    keys = get_configurable_keys()
    items = {k: v for k, v in items.items() if k in keys}
    # validate items - this will raise a validation Error a key is not valid
    try:
        Settings(**items)
    except ValidationError as exc:
        logging.getLogger(__name__).error('Invalid settings: %s', exc)
        raise exc

    with Session(connection) as session:
        session.expire_on_commit = False
        query = session.query(ConfigStore)
        current_conf = {u.config_key: u for u in query.all()}
        for key, value in items.items():
            if key not in current_conf:
                session.add(ConfigStore(config_key=key, config_val=value))
            elif current_conf[key].config_val != value:
                current_conf[key].config_val = value
                session.add(current_conf[key])

        session.commit()
    # Remove from environment variables not in the new brevia_env_secrets
    if 'brevia_env_secrets' in items:
        new_secrets = items['brevia_env_secrets']
        for key in get_settings().brevia_env_secrets:
            if key not in new_secrets:
                del environ[key]
    # Clear settings cache, force settings reload
    get_settings.cache_clear()

    return read_db_conf(connection)


def reset_db_conf(connection: Connection, items: list[str]) -> dict[str, str]:
    """ Reset to defatul some config items """
    # Filter out non-configurable settings
    keys = get_configurable_keys()
    items = [k for k in items if k in keys]
    with Session(connection) as session:
        session.expire_on_commit = False
        query = session.query(ConfigStore)
        current_conf = {u.config_key: u for u in query.all()}
        for key in items:
            if key in current_conf:
                session.delete(current_conf[key])
        session.commit()
    # Clear settings cache, force settings reload
    get_settings.cache_clear()

    return read_db_conf(connection)


@lru_cache
def get_settings():
    """Return Settings object instance just once (using lru_cache)"""
    settings = Settings()
    settings.setup_defaults()
    settings.update_from_db()
    settings.setup_environment()

    set_verbose(settings.verbose_mode)

    return settings
